textworld/generator/chaining.py

```diff
@@ -296,3 +296,211 @@ def print_chains(chains, verbose=False, backward=False):
+            action = self.try_instantiate(node.state, rule, mapping)
+            if not action:
+                continue
+
+            if not self.is_relevant(node, action):
+                continue
+
+            state = self.apply(node.state, action)
+            if not state:
+                continue
+
+            actions.append(action)
+            states.append(state)
+
+        for i, action in enumerate(actions):
+            remaining = actions[i+1:]
+            backtracks = node.backtracks + [remaining]
+            yield _ChainNode(node, states[i], action, backtracks, node.depth + 1, node.breadth)
+
+    def backtrack(self, node):
+        if node.breadth >= self.max_breadth:
+            return
+
+        for i, actions in enumerate(node.backtracks):
+            backtracks = node.backtracks[:i]
+
+            for j, action in enumerate(actions):
+                state = self.apply(node.state, action)
+                if not state:
+                    continue
+
+                remaining = actions[j+1:]
+                new_backtracks = backtracks + [remaining]
+                yield _ChainNode(node, state, action, new_backtracks, i + 1, node.breadth + 1)
+
+    def is_relevant(self, node, action):
+        if not node.action:
+            return True
+
+        if self.backward:
+            # XXX
+            post = action.removed
+            pre = node.action.postconditions
+        else:
+            post = node.action.added
+            pre = action.preconditions
+
+        return bool(set(post) & set(pre))
+
+    def apply(self, state, action):
+        new_state = state.copy()
+        for prop in action.preconditions:
+            if not new_state.is_fact(prop):
+                if all(state.has_variable(var) for var in prop.arguments):
+                    # Don't allow creating new predicates without any new variables
+                    return None
+                new_state.add_fact(prop)
+
+        # Make sure new_state still respects the constraints
+        if not self.check_state(new_state):
+            return None
+
+        new_state.apply(action)
+
+        # XXX: Some debug checks
+        assert self.check_state(new_state)
+
+        return new_state
+
+    def check_state(self, state):
+        fail = Proposition("fail", [])
+
+        constraints = state.all_applicable_actions(self.constraints)
+        for constraint in constraints:
+            if state.is_applicable(constraint):
+                # Optimistically delay copying the state
+                copy = state.copy()
+                copy.apply(constraint)
+
+                if copy.is_fact(fail):
+                    return False
+
+        return True
+
+    def try_instantiate(self, state, rule, mapping):
+        for ph in rule.placeholders:
+            if mapping.get(ph) is None:
+                # XXX: types_counts, max_types_counts
+                name = get_new(ph.type)
```

Comment:
`get_new()` needs `types_counts` which can be obtained from the state `types_counts = data.get_types().count(state)`. If you have a better of handling this notion of unique (but human-readable) variable names, I'm all ears.

Reply:
Yeah hence the comment above.  I want to migrate away from `data.get_types()` which is the old `VariableTypeTree` stuff and use the new logic types.

Reply:
Makes sense, great.